File: Data_Structure_Review/Queue_Implementation/linkedlist_queue.py
class ListNode:
    def __init__(self, val: int) -> None:
        self._val: int = val
        self._next: ListNode | None = None


class LinkedListQueue:
    """Linked list based queue implementation"""
    
    def __init__(self):
        """Constructor method"""
        self._head: ListNode | None = None # head node
        self._tail: ListNode | None = None # tail node
        self._size: int = 0
        
    def size(self) -> int:
        """Get the size of queue"""
        return self._size
    
    def is_empty(self) -> bool:
        """Check the queue is empty or not"""
        #return self.size() == 0
        return not self._head
    
    def push(self, val: int) -> None:
        """Push a new elt with val into queue"""
        # new a node to be ready to push
        new_node: ListNode = ListNode(val)
        # if the queue is empty, then let head, tail point to this new_node
        if self._head is None or self._tail is None:
            self._head = new_node
            self._tail = new_node
        # if not empty, then append this new_node after _tail
        else:
            self._tail._next = new_node
            self._tail = new_node
        self._size += 1
        
    def peek(self) -> int:
        """Access the head of queue and return its val"""
        if self._head is None:
            raise IndexError("Queue is empty!")
        return self._head._val
    
        
    def pop(self) -> int:
        """Pop an elt from the head and output its val"""
        num:int = self.peek()
        # delete the head node
        if self._head is None:
            raise IndexError("Queue is empty!")
        self._head = self._head._next
        self._size -= 1
        return num
    
    def to_list(self) -> list[int]:
        """Return a list version for printing"""
        queue:list[int] = []
        curr = self._head
        while curr:
            queue.append(curr._val)
            curr = curr._next
        return queue
            
if __name__ == "__main__":
    # 1️⃣  create a queue
    q = LinkedListQueue()
    print("Initially empty?", q.is_empty())       # True
    print("Initial size:", q.size())              # 0
    print("Queue state:", q.to_list())            # []

    # 2️⃣  enqueue some items
    for num in [10, 20, 30, 40]:
        q.push(num)
        print(f"After push({num}) →", q.to_list())

    # 3️⃣  inspect head without removing
    print("Peek head:", q.peek())                 # 10
    print("Current size:", q.size())              # 4

    # 4️⃣  dequeue two items
    print("Pop:", q.pop())                        # removes 10
    print("After pop:", q.to_list(), "OMFG")
    print("Pop:", q.pop())                        # removes 20
    print("After pop:", q.to_list())
    print("Current head:", q.peek())              # 30
    print("Current size:", q.size())              # 2

    # 5️⃣  enqueue again (tests tail update)
    q.push(50)
    print("After push(50):", q.to_list())         # [30, 40, 50]

    # 6️⃣  empty completely
    while not q.is_empty():
        print("Dequeued:", q.pop(), "→", q.to_list())

    print("Empty?", q.is_empty())                 # True
    print("Final size:", q.size())                # 0

File: Data_Structure_Review/Queue_Implementation/LinkedListQueue.java
package Queue_Implementation;

import java.util.NoSuchElementException;

/*
 * 	•	A static method belongs to the class itself, not any particular object.
	  •	A non-static (instance) method operates on the specific instance’s data (its fields like head, tail, size).
 */

public class LinkedListQueue {

  // private static inner class --- ListNode
  private static class ListNode {
    private int val;
    private ListNode next;

    // constructor
    public ListNode(int val){
        this.val = val;
        this.next = null;
    }
  }

  private ListNode head;
  private ListNode tail;
  private int size;

  // constructor of LinkedListQueue
  public LinkedListQueue(){
    this.head = null;
    this.tail = null;
    this.size = 0;
  }

  public int size() {
    return this.size;
  }

  public boolean isEmpty(){
    return this.head == null;
  }

  public void push(int val){
    //create new node with val
    ListNode newNode = new ListNode(val);
    if (isEmpty()){
      head = newNode;
      tail = newNode;
    } else {
      tail.next = newNode;
      tail = newNode;
    }
    size++;
  }

  public int peek(){
    if (isEmpty()){
      throw new NoSuchElementException("Queue is empty");
    }

    return head.val;
  }

  public int pop(){
    int headVal = peek();

    head = head.next;
    size--;
    return headVal;
  }


  @Override
  public String toString(){
    StringBuilder sb = new StringBuilder("[");
    ListNode curr = head;
    while (curr != null){
      sb.append(curr.val);
      if (curr.next != null) sb.append(", ");
      curr = curr.next;
    }
    sb.append("]");
    return sb.toString();
  }

  public static void main(String[] args) {
    LinkedListQueue q = new LinkedListQueue();
    System.out.println("Queue initially empty? " + q.isEmpty());
    System.out.println("Initial size: " + q.size());

    // Enqueue (push) some elements
    q.push(10);
    q.push(20);
    q.push(30);
    System.out.println("After pushing 10, 20, 30: " + q);
    System.out.println("Current size: " + q.size());

    // Peek at the front
    System.out.println("Front element (peek): " + q.peek());

    // Dequeue (pop) two elements
    System.out.println("Popped: " + q.pop());
    System.out.println("After first pop: " + q);
    System.out.println("Popped: " + q.pop());
    System.out.println("After second pop: " + q);
    System.out.println("Current size: " + q.size());

    // Push again to test tail updates
    q.push(40);
    q.push(50);
    System.out.println("After pushing 40, 50: " + q);

    // Empty the queue completely
    while (!q.isEmpty()) {
        System.out.println("Dequeued " + q.pop() + " → " + q);
    }

    System.out.println("Queue empty? " + q.isEmpty());
  }
}

